packages/chain-mon/src/faultproof-wd-mon/service.ts

```diff
@@ -0,0 +1,519 @@
+    let latestL1BlockNumber: number
+    try {
+      latestL1BlockNumber = await this.options.l1RpcProvider.getBlockNumber()
+    } catch (err) {
+      // Log the issue so we can debug it.
+      this.logger.error(`got error when connecting to node`, {
+        error: err,
+        node: 'l1',
+        section: 'getBlockNumber',
+      })
+
+      // Increment the metric so we can detect the issue.
+      this.metrics.nodeConnectionFailures.inc({
+        layer: 'l1',
+        section: 'getBlockNumber',
+      })
+
+      // Sleep for a little to give intermittent errors a chance to recover.
+      return sleep(this.options.sleepTimeMs)
+    }
+
+    // Update highest block number metrics so we can keep track of how the service is doing.
+    this.metrics.highestKnownBlockNumber.set(latestL1BlockNumber)
+    this.metrics.highestCheckedBlockNumber.set(
+      this.state.highestUncheckedBlockNumber
+    )
+
+    // Check if the RPC provider is behind us for some reason. Can happen occasionally,
+    // particularly if connected to an RPC provider that load balances over multiple nodes that
+    // might not be perfectly in sync.
+    if (latestL1BlockNumber <= this.state.highestUncheckedBlockNumber) {
+      // Sleep for a little to give the RPC a chance to catch up.
+      return sleep(this.options.sleepTimeMs)
+    }
+
+    // Generally better to use a relatively small block range because it means this service can be
+    // used alongside many different types of L1 nodes. For instance, Geth will typically only
+    // support a block range of 2000 blocks out of the box.
+    const toBlockNumber = Math.min(
+      this.state.highestUncheckedBlockNumber + this.options.eventBlockRange,
+      latestL1BlockNumber
+    )
+
+    // Useful to log this stuff just in case we get stuck or something.
+    this.logger.info(`checking recent blocks`, {
+      fromBlockNumber: this.state.highestUncheckedBlockNumber,
+      toBlockNumber,
+    })
+
+    // Query for WithdrawalProven events within the specified block range.
+    let events: ethers.Event[]
+    try {
+      events = await this.state.portal.queryFilter(
+        this.state.portal.filters.WithdrawalProven(),
+        this.state.highestUncheckedBlockNumber,
+        toBlockNumber
+      )
+    } catch (err) {
+      // Log the issue so we can debug it.
+      this.logger.error(`got error when connecting to node`, {
+        error: err,
+        node: 'l1',
+        section: 'querying for WithdrawalProven events',
+      })
+
+      // Increment the metric so we can detect the issue.
+      this.metrics.nodeConnectionFailures.inc({
+        layer: 'l1',
+        section: 'querying for WithdrawalProven events',
+      })
+
+      // Sleep for a little to give intermittent errors a chance to recover.
+      return sleep(this.options.sleepTimeMs)
+    }
+
+    // Go over all the events and check if the withdrawal hash actually exists on L2.
+    for (const event of events) {
```

Comment:
If this loop throws for whatever reason then the same event may be dropped into `invalidProposalWithdrawals` or `invalidProofWithdrawals` more than once. I don't think this ends up being an issue but it can lead to inflated metrics. Personally I would rather have inflated metrics than not increment the metrics at all -- probably just worth documenting this behavior.